Automattic/jetpack
# Search package: allow auto config add search input for block themes

Fixes #22611

#### Changes proposed in this Pull Request:
Added functionality to add search input for block-based themes. 

<strike>The search input is added above footer and on the 3rd columns of three. If you think there's a better place to put it, let me know. 😄 As suggested by @gibrown, I moved the search block just below header:</strike>

After some internal and external discussions (please refer to the comments of the PR), we decided to add the search block in the footer. For one it's pretty obvious to notice and for two it doesn't disturb users too much.

- Not to add search block if there's already search block(s).

- Try to get the content of footer template and parse it. 
    - If there are columns in the footer, the code would try to append the search block to the first column.

![image](https://user-images.githubusercontent.com/1425433/153357029-9639d15a-2650-4266-a0a3-f260954a4789.png)

    - If there's no columns, the code would try to add the search block to the inner group.

![image](https://user-images.githubusercontent.com/1425433/153357308-5a63010a-a694-4dc6-9544-2d00562e044d.png)

- Some templates has block patterns in templates, the code would replace them before doing changes described above.

- If there are not groups or columns then the template is not touched.

Note: the replacement is not recursive, considering to minimize the risk. The `auto_config` command is updated to work with the patch.
 

#### Jetpack product discussion
pcNPJE-Cx-p2
p1644361560289179-jetpack-search-backstage-slack

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
##### Test on JN
* Create a JN site with WP 5.9 and apply a block theme .e.g. [twentytwentytwo](https://wordpress.org/themes/twentytwentytwo/) (allows FSE)
* Purchase JP search product
* Ensure a search input is added in footer and looks reasonable
* Ensure it summons search overlay on input

#### Test on your local
- Ensure you have a instant search enabled
- Switch to a clean block theme ( I tested twentytwentytwo, alara, stewart )
- Run command `jetpack docker wp jetpack-search auto_config`
- Ensure search block is added in footer and looks reasonable
- Ensure PHP tests pass: `( cd projects/packages/search && composer test-php )`
- Please test more block themes if you could

##### how to test on JN
- Open JN special Ops, create a site with the following parameters:
![image](https://user-images.githubusercontent.com/1425433/152453393-76d163ce-71b3-41bd-8917-7cb6b8e75990.png)
- Connect the site with a WPCOM user
- Purchase a JP Search plan
